sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,288 @@
+        k_dist = distances[:, self.n_neighbors-1]
+
+        return k_dist, neighbors_indices
+
+    def _local_reachability_density(self, p):
+        """The local reachability density (LRD) of p is the inverse of the
+        average reachability distance based on the self.n_neighbors-nearest
+        neighbors of instance.
+
+        Parameters
+        ----------
+        p : array-like of shape (n_samples, n_features)
+        The samples to compute the LRD w.r.t. self._fit_X
+        If None, compute the LRD of self._fit_X w.r.t. to itself.
+        (In this case samples are not considered in their own neighbourhood)
+
+        Returns
+        -------
+        local reachability density : float
+        The LRD of p.
+        """
+
+        p_0 = self._fit_X if p is None else p
+
+        neighbors_indices = self.neighbors_indices_fit_X_ if p is None else self._k_distance(p)[1]
+
+        # for efficiency, use squared euclidean distances
+        if self.effective_metric_ == 'euclidean':
+            dist = pairwise_distances(p_0, self._fit_X, 'euclidean',
+                                      n_jobs=self.get_n_jobs_, squared=True)
+        else:
+            dist = pairwise_distances(
+                p_0, self._fit_X, self.effective_metric_, n_jobs=self.get_n_jobs_,
+                **self.effective_metric_params_)
+
+        reach_dist_array = np.zeros((p_0.shape[0], self.n_neighbors))
+
+        for j in range(p_0.shape[0]):
+            neighbors_number = -1
+            for i in neighbors_indices[j, :]:
+                neighbors_number += 1
+                if self.effective_metric_ == 'euclidean':
+                    reach_dist_array[j, neighbors_number] = np.max(
+                        [self._k_distance_value_fit_X_[i],  np.sqrt(dist[j, i])])
+                else:
+                    reach_dist_array[j, neighbors_number] = np.max(
+                        [self._k_distance_value_fit_X_[i],  dist[j, i]])
+
+        return self.n_neighbors / np.sum(reach_dist_array, axis=1)
+
+    def _local_outlier_factor(self, p=None):
+        """ The (local) outlier factor (LOF) of instance captures its
+        supposed `degree of abnormality'.
+        It is the average of the ratio of the local reachability density of p
+        and those of p's self.n_neighbors-NN.
+
+        Parameters
+        ----------
+        p : array-like of shape (n_samples, n_features)
+        The points to compute the LOF w.r.t. training samples X=self._fit_X.
+        Note that samples p are not considered in the neighbourhood of X for
+        computing local_reachability_density of samples X.
+        If None, compute LOF of self._fit_X w.r.t. to itself. (In this case
+        samples are not considered in their own neighbourhood)
+
+        Returns
+        -------
+        Local Outlier Factor : array-like of shape (n_samples,)
+        The LOF of p. The lower, the more normal.
+
+        """
+        p_0 = self._fit_X if p is None else p
```

Comment:
We only keep X as attribute when `p is None`, ie when the querry samples are equal to the training ones (in this case indeed, the base class method `KNeighborsMixin.kneighbors` checks if min dist is 0 and removes the first sample in this case). I believe this case is optimal...

If the querry samples are different from the training ones, `p is not None` and `p` contains the querry samples. As commented the line below, this case is not really optimal for now, since only the neighbors of neighbors of `p` need to be considered (LOF is local).